### PR TITLE
[EDU-1654] - Add generate Ably token authentication example

### PR DESCRIPTION
**Pull request by GregHolmes**
This PR contains the two code samples for the new Examples section we are creating. It contains a Typescript and a React version for authentication requesting an access token from a backend server. This is a different structure example to most. This one requires a server to be running to generate the token so I've created an example for this in the server directory.

The README within each directory `/examples/auth-generate-token/javascript` and `/examples/auth-generate-token/react` contains information on how to get these running.

[Jira epic ticket](https://ably.atlassian.net/browse/EDU-1541)
[Jira ticket](https://ably.atlassian.net/browse/EDU-1654).

### CODE REVIEW

File: examples/auth-generate-token/page.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+This example uses a frontend and a backend client. The frontend is a Typescript application and the backend is an Express Node.js application. Both of these applications use the [Ably JS SDK](https://github.com/ably/ably-js).
+
+The server application calls [`auth.requestToken();`](https://ably.com/docs/auth/token#ably-token) to request a token from Ably in order to pass it to the client to be used as authentication. The client calls the server's publicly exposed endpoint `/auth-url` to request this token.
+
+Find out more about [authentication](https://ably.com/docs/auth).
+
+// End Javascript brief
+
+## View on Github
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:33]: You need to add an additional step to these instructions to cover adding your API key to the server.

File: examples/auth-generate-token/page.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
+# Generate a JWT token to authenticate clients
+
+Use JWT tokens to securely authenticate users within your application.
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:3]: "JWT tokens" or "JWTs"?

File: examples/auth-generate-token/page.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+JWT tokens enable you to grant users access to Ably's real-time messaging services. They are most commonly used to issue time-limited credentials that securely authorize users to connect to Ably channels.
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:5]:
```suggestion
JWT tokens enable you to grant users access to the Ably platform. They are used to issue time-limited credentials that securely authorize users, and manage which resources each user has access to.
```

File: examples/auth-generate-token/page.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+JWT tokens are an essential component of applications that require secure and scalable real-time communication. Before any application can use any of the Ably products whether they be Pub/Sub, Spaces, Chat, or LiveSync, they'll need to first authenticate.
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:7]:
```suggestion
The first step for an application to communicate with Ably requires them authenticating.
```

I think we should replace the deleted info here with explaining about authUrl and token renewal. WDYT?

File: examples/auth-generate-token/page.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+Authentication is implemented using [Ably core Pub/Sub](https://ably.com/docs/auth). The Pub/Sub SDK contains a set of purpose-built APIs that abstract away the complexities involved in architecting websocket features. It is built on top of Ably's core platform, and so it provides the same performance guarantees and scaling potential.
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:9]: I think we need to reword this to state that authentication is handled by Pub/Sub and how that bleeds into the other products.

File: examples/auth-generate-token/page.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+## Resources
+
+// React brief (Only visible if viewing the React example)
+
+This example uses a frontend and a backend client. The frontend is a React application and the backend is an Express Node.js application. Both of these applications use the [Ably JS SDK](https://github.com/ably/ably-js).
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:15]:
```suggestion
This example uses a frontend and a backend client. The frontend is a React application and the backend is an Express Node.js application. Both of these applications use Ably Pub/Sub.
```

File: examples/auth-generate-token/page.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+The server application calls [`auth.requestToken();`](https://ably.com/docs/auth/token#ably-token) to request a token from Ably in order to pass it to the client to be used as authentication. The client calls the server's publicly exposed endpoint `/auth-url` to request this token.
+
+Find out more about [authentication](https://ably.com/docs/auth).
+
+// End React brief
+
+// Javascript brief (Only visible if viewing the Javascript example)
+
+This example uses a frontend and a backend client. The frontend is a Typescript application and the backend is an Express Node.js application. Both of these applications use the [Ably JS SDK](https://github.com/ably/ably-js).
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:25]:
```suggestion
This example uses a frontend and a backend client. The frontend is a Typescript application and the backend is an Express Node.js application. Both of these applications use Ably Pub/Sub.
```

**Review by m-hulbert — CHANGES_REQUESTED** [review]
This is going to be really useful, couple of suggestions: 

I'd suggest that we have the steps already printed out, and then as they occur they receive a tick at the end of line. I think this might show the speed a bit more clearly, and also avoid the fact you can keep pushing 'connect'. 

I know we spoke about it previously, but I think it would be worth expanding on all the steps and adding in everything that happens, even if some of it is server-side. 

```
1. Client requests token from server
2. Client initializes connection to Ably with generated Token
```

Minor point: I think having a title like "Welcome to the Ably Token Manager" makes the process seem a bit more grandiose, when it's supposed to be (is) quite simple. Thoughts?

File: examples/auth-generate-token/page.md (hunk 1 above)
Comment (GregHolmes) [reply examples/auth-generate-token/page.md:33]: I'm sure I added a comment here but I can't find it... Steps 5-10 are these steps.

File: examples/auth-generate-token/page.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
+# Generate an Ably token to authenticate clients
+
+Use an Ably token to securely authenticate users within your application.
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:3]:
```suggestion
Use an Ably token to securely authenticate your users.
```

File: examples/auth-generate-token/page.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
+# Generate an Ably token to authenticate clients
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:1]: Should this be "request" rather than "generate" (and generate is for JWT?)

Also "Ably Token" needs to be capitalised throughout (sorry)

File: examples/auth-generate-token/page.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+Use an Ably token to securely authenticate users within your application.
+
+Ably tokens enable you to grant users access to the Ably platform. They are used to issue time-limited credentials that securely authorize users, and manage which resources each user has access to.
+
+Authentication is the first step for an application to communicate with Ably. The application initializes a client and calls a predefined `authUrl` endpoint on your backend server. This endpoint uses an API key to request a token from Ably and returns it to the requesting client. When the token nears expiry, the `authUrl` is automatically invoked to request a new token.
+
+Authentication is implemented using [Ably core Pub/Sub](https://ably.com/docs/auth). The Pub/Sub SDK provides authentication capabilities that are shared across all Ably products. The SDK includes purpose-built APIs that streamline pub/sub features while delivering Ably's performance guarantees and scaling capabilities.
```
Comment (m-hulbert) [inline examples/auth-generate-token/page.md:9]:
```suggestion
Authentication is implemented using [Ably core Pub/Sub](https://ably.com/docs/auth). The Pub/Sub SDK provides the authentication mechanism that is utilized by all Ably products.
```

Then finish it off with whatever line we agree on regarding the Pub/Sub SDK (`The Pub/Sub SDK provides a set of flexible APIs capable of building any realtime application. It is powered by Ably's reliable and scalable platform.`) was my suggestion in the other PR.

**Review by m-hulbert — APPROVED** [review]
Minor: should the directory be "auth-request-token" rather than generate? 

Minor suggestions but LGTM 🙂
